### PR TITLE
[10.0][MIG] account_asset_disposal

### PR DESCRIPTION
**Pull request by luismontalba**
Account asset disposal
-----------------------------
This module extends the functionality of account_asset adding a disposal
date and allowing reversion of disposal operation.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Wainting for a proper rebase.

File: account_asset_disposal/README.rst [hunk 1]
```diff
@@ -1,28 +1,23 @@
...
+   :alt: License: AGPL-3
 
 ======================
 Account asset disposal
 ======================
 
-This module extends the functionality of account_asset and automatically
-generates the asset close move
+This module extends the functionality of account_asset adding a disposal
+date and allowing reversion of disposal operation.
```
Comment (pedrobaeza) [inline account_asset_disposal/README.rst]: What about the new state?

File: account_asset_disposal/models/__init__.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from . import account_asset_category
 from . import account_asset_asset
 from . import account_move_line
+from . import account_asset_depreciation_line
```
Comment (pedrobaeza) [inline account_asset_disposal/models/__init__.py]: Why don't you include this one in `account_asset_asset` as it's a one2many submodel?

File: account_asset_disposal/models/account_asset_asset.py [hunk 3]
```diff
@@ -102,6 +32,15 @@ def action_disposal(self):
         }
 
     @api.multi
+    def set_to_close(self):
+        res = super(AccountAssetAsset, self).set_to_close()
+        if res:
+            self.disposal_move_id = self.env['account.move'].search([
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_asset_asset.py]: Put directly `self.disposal_move_id = res['res_id']`

File: account_asset_disposal/models/account_asset_asset.py [hunk 4]
```diff
@@ -102,6 +32,15 @@ def action_disposal(self):
...
+                ('id', '=', res['res_id'])])
+        return res
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_asset_asset.py]: You are not changing state here to `disposed`.

File: account_asset_disposal/models/account_asset_asset.py [hunk 5]
```diff
@@ -111,6 +50,7 @@ def action_disposal_undo(self):
                 asset.state = 'close'
             else:
                 asset.state = 'open'
+                asset.method_number = asset.category_id.method_number
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_asset_asset.py]: Why this?

File: account_asset_disposal/models/account_asset_asset.py [hunk 6]
```diff
@@ -111,6 +50,7 @@ def action_disposal_undo(self):
...
         return self.write({
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_asset_asset.py]: You are not recalculating depreciation board.

File: account_asset_disposal/models/account_asset_depreciation_line.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, models
+
+class AccountAssetDepreciationLine(models.Model):
+    _inherit = "account.asset.depreciation.line"
+
+
+    @api.multi
+    def post_lines_and_close_asset(self):
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_asset_depreciation_line.py]: Never overwrite methods without super. And why is this needed?

File: account_asset_disposal/tests/test_account_asset_disposal.py [hunk 8]
```diff
@@ -100,7 +101,7 @@ def setUp(self):
             ('special', '=', False),
         ], limit=1, order='date_stop DESC')
 
-    def test_asset_disposal(self):
+    # def test_asset_disposal(self):
```
Comment (pedrobaeza) [inline account_asset_disposal/tests/test_account_asset_disposal.py]: Don't comment tests

File: account_asset_disposal/views/account_asset_asset_view.xml [hunk 9]
```diff
@@ -1,54 +1,41 @@
...
+        <field name="arch" type="xml">
+            <button name="set_to_close" position="attributes">
+                <attribute name='invisible'>1</attribute>
+            </button>
+            <button name="compute_depreciation_board" position="attributes">
+                <attribute name="attrs">{
+                    'invisible':[('state', '!=', 'draft')]}</attribute>
+            </button>
+            <button name="set_to_close" position="after">
+                <button name="action_disposal"
```
Comment (pedrobaeza) [inline account_asset_disposal/views/account_asset_asset_view.xml]: Why this new button? Use the current one

File: account_asset_disposal/wizards/account_asset_disposal_wizard.py [hunk 10]
```diff
@@ -11,27 +12,18 @@ class AccountAssetDisposalWizard(models.TransientModel):
     def _default_disposal_date(self):
```
Comment (pedrobaeza) [inline account_asset_disposal/wizards/account_asset_disposal_wizard.py]: You don't need this wizard anymore

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: account_asset_disposal/README.rst (hunk 1 above)
Comment (luismontalba) [reply account_asset_disposal/README.rst]: I'll include explanation in README

File: account_asset_disposal/models/__init__.py (hunk 2 above)
Comment (luismontalba) [reply account_asset_disposal/models/__init__.py]: This is not needed now

File: account_asset_disposal/models/account_asset_asset.py (hunk 3 above)
Comment (luismontalba) [reply account_asset_disposal/models/account_asset_asset.py]: Done

File: account_asset_disposal/models/account_asset_asset.py (hunk 4 above)
Comment (luismontalba) [reply account_asset_disposal/models/account_asset_asset.py]: I's done it in action_disposal method

File: account_asset_disposal/models/account_asset_asset.py (hunk 5 above)
Comment (luismontalba) [reply account_asset_disposal/models/account_asset_asset.py]: To recalculate depreciation board according to number of entries set in the asset category.

File: account_asset_disposal/models/account_asset_asset.py (hunk 6 above)
Comment (luismontalba) [reply account_asset_disposal/models/account_asset_asset.py]: This way auto-recalculates the depreciation board according to the number of entries set in the asset category.

File: account_asset_disposal/models/account_asset_depreciation_line.py (hunk 7 above)
Comment (luismontalba) [reply account_asset_disposal/models/account_asset_depreciation_line.py]: This is not needed now

File: account_asset_disposal/tests/test_account_asset_disposal.py (hunk 8 above)
Comment (luismontalba) [reply account_asset_disposal/tests/test_account_asset_disposal.py]: Test are not migrated yet, WIP

File: account_asset_disposal/views/account_asset_asset_view.xml (hunk 9 above)
Comment (luismontalba) [reply account_asset_disposal/views/account_asset_asset_view.xml]: The new button is calling another function and is positioned in the view related to the old button.

File: account_asset_disposal/wizards/account_asset_disposal_wizard.py (hunk 10 above)
Comment (luismontalba) [reply account_asset_disposal/wizards/account_asset_disposal_wizard.py]: Wizard removed

File: account_asset_disposal/models/account_asset_asset.py (hunk 5 above)
Comment (pedrobaeza) [reply account_asset_disposal/models/account_asset_asset.py]: Yes, but you can't change the currrent number. It can be customize for each asset. The value set is already correct.

File: account_asset_disposal/views/account_asset_asset_view.xml (hunk 9 above)
Comment (pedrobaeza) [reply account_asset_disposal/views/account_asset_asset_view.xml]: Why can't you call the original method directly?

**Comment by luismontalba** [comment]
I think it's what I have done

**Comment by pedrobaeza** [comment]
No, it isn't. Just remove the depreciation line part.

File: account_asset_disposal/migrations/10.0.1.0.0/pre-migration.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    openupgrade.update_module_names(
+        cr, [
+            ('account_asset_disposal_analytic', 'account_asset_analytic'),
+        ], merge_modules=True,
+    )
+
+
+@openupgrade.migrate()
+def migrate(cr, version):
+    openupgrade.update_module_names(
```
Comment (pedrobaeza) [inline account_asset_disposal/migrations/10.0.1.0.0/pre-migration.py]: Remove this instruction which clearly fails.

File: account_asset_disposal/models/account_asset_asset.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    def action_disposal_undo(self):
+        for asset in self.with_context(asset_disposal_undo=True):
+            if asset.disposal_move_id:
+                asset.disposal_move_id.button_cancel()
+                asset.disposal_move_id.unlink()
+            if asset.currency_id.is_zero(asset.value_residual):
+                asset.state = 'close'
+            else:
+                asset.state = 'open'
+                asset.method_number = asset.category_id.method_number
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_asset_asset.py]: Why assign number of depreciations instead of calling `compute_depreciation_board`? Take into account that assets can be depreciated other way (in Spain with l10n_es_account_asset, with a percentage), so this won't work with that method.

File: account_asset_disposal/models/account_move_line.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+from odoo import _, api, models
+from odoo.exceptions import Warning as UserError
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    @api.multi
+    def unlink(self):
+        disposed_lines = self.env['account.asset.asset'].search([]).filtered(
```
Comment (pedrobaeza) [inline account_asset_disposal/models/account_move_line.py]: You shouldn't check this (as this is very performance drainer), and disposal move should be treated the same as the rest of the moves created from asset (it's inserted also in the depreciation board): if Odoo allows to manually remove them, then let it do it.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by luismontalba** [comment]
@pedrobaeza, changes proposed done and test added.

**Review by pedrobaeza — APPROVED** [review]
@rafaelbn please review

**Comment by pedrobaeza** [comment]
@cubells please review

File: account_asset_disposal/README.rst [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+   :target: http://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+======================
+Account asset disposal
+======================
+
+This module extends the functionality of account_asset adding a disposal
+date, allowing reversion of disposal operation and adding the state
+"Dispoded" to the asset.
```
Comment (cubells) [inline account_asset_disposal/README.rst]: s/Dispoded/Disposed/

File: account_asset_disposal/__manifest__.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    "category": "Accounting & Finance",
+    "website": "https://www.tecnativa.com",
+    "author": "Tecnativa, "
+              "Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "depends": [
+        "account_asset",
+        "account_cancel"
```
Comment (cubells) [inline account_asset_disposal/__manifest__.py]: End with comma, please.

**Review by cubells — APPROVED** [review]
Minor changes.

Code review and test.

LGTM

**Comment by luismontalba** [comment]
Changes proposed done.

**Comment by pedrobaeza** [comment]
@luismontalba squash for merging this.

**Comment by luismontalba** [comment]
Commits squashed.
